base/dict.jl

```diff
@@ -821,6 +844,39 @@ function get(default::Callable, dict::ImmutableDict, key)
     return default()
 end
 
+function delete(d::ImmutableDict{K,V}, key) where {K,V}
+    if isdefined(d, :parent)
+        if isequal(d.key, key)
+            d.parent
+        else
+            ImmutableDict{K,V}(delete(d.parent, key), d.key, d.value)
+        end
+    else
+        d
+    end
+end
```

Comment:
This implementation is not quite correct since I can current;y construct an ImmutableDict where there are multiple entries for `key` each shadowed.

Reply:
I suspect `delete` would be more easily done with adding a new key as a tombstone, rather than a bulk copy like this.

Reply:
> I suspect `delete` would be more easily done with adding a new key as a tombstone (with the same `key`, but `val` being `#undef`), rather than a bulk copy like this.

I really like that idea, but I think we would need to have a new type since we can't rely on `#undef` for bits types.